libs/net/kafka/KafkaListener`.cs

```diff
@@ -193,20 +217,56 @@ public async Task ConsumeAsync(Func<ConsumeResult<TKey, TValue>, Task<ConsumerAc
         {
-            // https://github.com/edenhill/librdkafka/blob/master/INTRODUCTION.md#fatal-consumer-errors
-            // An unhandled exception will stop consuming as I don't know of a way to resume the paused consumer.
-            _logger.LogError(ex, "Error while consuming. {message}", ex.Message);
-            proceed = this.OnError?.Invoke(this, new ErrorEventArgs(ex)) ?? ConsumerAction.Stop;
+            proceed = HandleException(ex);
         }
         finally
         {
-            if (proceed == ConsumerAction.Stop) this.Stop();
-            else if (this.IsPaused && proceed == ConsumerAction.Proceed)
+            if (!IsTranscription) FinalCleanUp(proceed);
+        }
+    }
+
+    private async Task<ConsumerAction> RunActionAsync(
+        Func<ConsumeResult<TKey, TValue>, Task<ConsumerAction>> action,
+        ConsumeResult<TKey, TValue>? _currentResult)
+    {
+        var result = await action(_currentResult!);
+        if (result == ConsumerAction.Proceed)
+        {
+            // Manually commit to inform Kafka this message has successfully been handled.
+            // TODO: Convert to more efficient process and use EnableAutoOffsetStore=false - https://docs.confluent.io/kafka-clients/dotnet/current/overview.html#store-offsets
+            if (_config.EnableAutoCommit == false)
+            {
+                this.Consumer?.Commit(_currentResult);
+                _logger.LogDebug("Message committed from Kafka topic:'{topic}' key:'{key}'", _currentResult!.Topic, _currentResult.Message.Key);
+            }
+            
+            if (IsTranscription)
             {
-                _logger.LogDebug("Resuming consumption: {topics}", String.Join(", ", this.Consumer!.Subscription ?? new List<string>()));
-                this.IsPaused = false;
-                this.Consumer?.Resume(this.Consumer.Assignment);
+                Interlocked.Decrement(ref _currentThread); 
+                _currentResults.TryRemove(_currentResult!, out _);
             }
+
+            _currentResult = null;
+        }
+        return result;
+    }
+
+    private ConsumerAction HandleException(Exception ex)
+    {
+        // https://github.com/edenhill/librdkafka/blob/master/INTRODUCTION.md#fatal-consumer-errors
+        // An unhandled exception will stop consuming as I don't know of a way to resume the paused consumer.
+        _logger.LogError(ex, "Error while consuming. {message}", ex.Message);
+        return this.OnError?.Invoke(this, new ErrorEventArgs(ex)) ?? ConsumerAction.Stop;
+    }
+
+    private void FinalCleanUp(ConsumerAction proceed)
+    {
+        if (proceed == ConsumerAction.Stop) this.Stop();
+        else if (this.IsPaused && proceed == ConsumerAction.Proceed)
+        {
+            _logger.LogDebug("Resuming consumption: {topics}", String.Join(", ", this.Consumer!.Subscription ?? new List<string>()));
+            this.IsPaused = false;
+            this.Consumer?.Resume(this.Consumer.Assignment);
```

Comment:
this will not work correctly within your threads.  The `this.Consumer.Assignment` will be the most recent result message from Kafka.

Reply:
This seems to be working fine, or maybe I miss something else?

Reply:
Consider the following; message 1 received. Sends message 1 to get transcription.  While transcribing, message 2 is received.  Sends message 2 to get transcription.  Message 2 completes first.  You are now telling Kafka what exactly?  Are you telling Kafka that only message 2 has been completed?  Or are you telling Kafka both message 1 and 2 are completed?  Now what happens when message 1 completes.  What is the value of `this.Consumer.Assignment` at that point in time?
I doubt this works reliably.

Reply:
Pause only occurs
a) when `UseMultiThreads == false`
b) when `ReachOrOverLimit == true`

The value of `proceed` may have been `ConsumerAction.Proceed` for some of the transcriptions but not all when `UseMultiThreads == true`

The value of `_currentResults.IsEmpty == true` only when all threads have been completed and no new messages have been received from Kafka.

This means that this won't work for multi-threading.

You need to inform Kafka that when a message has been successfully handled that it should consider it completed so that it is not sent to get transcribed again.  Only failures should result in Kafka resending the same message.

Reply:
Code updated, and even though the value of proceed could be something else than ConsumerAction.Proceed, however, they could only have 2 choices either ConsumerAction.Retry or ConsumerAction.Stop, and eventually Retry could become Proceed or Stop as well, so in this case it should work for multi-threading. Thanks

Reply:
What is the value of `this.Consumer.Assignment` when the following code executes `this.Consumer?.Resume(this.Consumer.Assignment)`?  

It needs to be the specific message assignment that was successful.  Presently it has no way to control this.